be/src/vec/functions/round.h

```diff
@@ -446,6 +479,191 @@ struct Dispatcher {
+            Int32 input_scale = data_col_general.get_scale();
+            const size_t input_rows_count = col_scale->size();
+            auto col_res = ColumnDecimal<T>::create(input_rows_count, input_scale);
+
+            for (size_t i = 0; i < input_rows_count; ++i) {
+                const Int32 scale_arg = col_scale_i32.get_data()[i];
+
+                if (scale_arg > std::numeric_limits<Int16>::max() ||
+                    scale_arg < std::numeric_limits<Int16>::min()) {
+                    throw doris::Exception(ErrorCode::OUT_OF_BOUND,
+                                           "Scale argument for function is out of bound: {}",
+                                           scale_arg);
+                }
+            }
+
+            for (size_t i = 0; i < input_rows_count; ++i) {
+                DecimalRoundingImpl<T, rounding_mode, tie_breaking_mode>::apply(
+                        general_val, input_scale, col_res->get_element(i).value,
+                        col_scale_i32.get_data()[i]);
+            }
+
+            for (size_t i = 0; i < input_rows_count; ++i) {
+                // For truncate(ColumnDecimal, ColumnInt32), we should always have same scale with source Decimal column
+                // So we need this check to make sure the result have correct digits count
+                //
+                // Case 0: scale_arg <= -(integer part digits count)
+                //      do nothing, because result is 0
+                // Case 1: scale_arg <= 0 && scale_arg > -(integer part digits count)
+                //      decimal parts has been erased, so add them back by multiply 10^(scale_arg)
+                // Case 2: scale_arg > 0 && scale_arg < decimal part digits count
+                //      decimal part now has scale_arg digits, so multiply 10^(input_scale - scal_arg)
+                // Case 3: scale_arg >= input_scale
+                //      do nothing
+                const Int32 scale_arg = col_scale_i32.get_data()[i];
+                if (scale_arg <= 0) {
+                    col_res->get_element(i).value *= int_exp10(input_scale);
+                } else if (scale_arg > 0 && scale_arg < input_scale) {
+                    col_res->get_element(i).value *= int_exp10(input_scale - scale_arg);
+                }
+            }
+
+            return col_res;
+        } else if constexpr (IsNumber<T>) {
+            const ColumnVector<T>& data_col_general =
+                    assert_cast<const ColumnVector<T>&>(const_col_general->get_data_column());
+            T general_val = data_col_general.get_data()[0];
+            const size_t intput_rows_cound = col_scale->size();
+            auto col_res = ColumnVector<T>::create(intput_rows_cound);
+            typename ColumnVector<T>::Container& vec_res = col_res->get_data();
+
+            for (size_t i = 0; i < intput_rows_cound; ++i) {
+                const Int16 scale_arg = col_scale_i32.get_data()[i];
```

Comment:
in apply_vec_vec function, have check all data of scale_arg in [min, max]
here seems not have this check, it's no need or in other function?

Reply:
Yes we need to do the check. I will re-organize the check loop